### PR TITLE
Fix `phasor_calibrate` to handle multi harmonic calibration

### PR DESCRIPTION
**Pull request by bruno-pannunzio**
## Description
This PR fixes the issue of `phasor_calibrate` being unable to handle correctly  multiple harmonic calibration. This is related to issue #68.

Not related, but also fixed a minor issue related to #67, to allow `phasor_center` to pass other keyword arguments other than `axis` to numpy functions used in methods. 
...

## Release note

Summarize the changes in the code block below to be included in the
[release notes](https://www.phasorpy.org/stable/release.html):

```release-note
Fix `phasor_calibrate` to handle multiple-harmonic calibration.
```

## Checklist

- [x] The pull request title, summary, and description are concise.
- [x] Related issues are linked in the description.
- [x] New dependencies are explained.
- [x] The source code and documentation can be distributed under the [MIT license](https://www.phasorpy.org/stable/license.html).
- [x] The source code adheres to [code standards](https://www.phasorpy.org/stable/contributing.html#code-standards).
- [x] New classes, functions, and features are thoroughly [tested](https://www.phasorpy.org/stable/contributing.html#tests).
- [x] New, user-facing classes, functions, and features are [documented](https://www.phasorpy.org/stable/contributing.html#documentation).
- [x] New features are covered in tutorials.
- [x] No files other than source code, documentation, and project settings are added to the repository.

### CODE REVIEW

**Comment by bruno-pannunzio** [comment]
@cgohlke, I implemented some changes to `phasor_transform` and `phasor_calibrate` in order to properly handle calibration of multiple harmonics, as stated in issue #69. If you consider this is not necessary, we can close this PR, but I was unable to calibrate multiple harmonics of real data with previous implementations, at least with the pipeline I posted. I also added to the `phasorpy_introduction` tutorial an example for calibration of multiple harmonics. If you think it is not necessary I can remove it.

Finally, I modified the `phasor_center` function to allow passing other keyword arguments to numpy functions other than `axis`. I don't think this is necessary for the moment, but I think it may be useful later, although I can't think of an exact use right now. Also, if you believe this is not necessary, I can remove it.

**Comment by cgohlke** [comment]
> I implemented some changes to phasor_transform and phasor_calibrate in order to properly handle calibration of multiple harmonics, as stated in issue #69. If you consider this is not necessary, we can close this PR

Yes, I think this case should be supported by the `phasor_calibrate` function if possible. My objection is to "fix" the `phasor_transform` function. Being a numpy ufunc, it is already capable to broadcast in the most general way. I suggest to move your changes out of `phasor_transform` into `phasor_calibrate`. That should work AFAICT. No?

> I also added to the phasorpy_introduction tutorial an example for calibration of multiple harmonics. If you think it is not necessary I can remove it.

Right now it looks out of place. However, if you think the tutorial will make use of the second harmonic later, for example, to calculate multiple components, this section should replace the single harmonic calibration.

> I modified the phasor_center function to allow passing other keyword arguments to numpy functions other than axis

Good idea. The one issue I can see is passing through the `out` parameter. To support `out`  in the `phasor_center` function, it should be a tuple `(out_real, out_imag)`.

File: src/phasorpy/phasor.py [hunk 1]
```diff
@@ -684,14 +689,32 @@ def phasor_transform(
...
+    dim_diff = numpy.ndim(re) - numpy.ndim(phase_zero)
+    if dim_diff > 0:
+        phase_zero = numpy.expand_dims(
+            phase_zero,
+            axis=tuple(i for i in range(-dim_diff, 0) if i not in skip_axes),
+        )
+        modulation_zero = numpy.expand_dims(
+            modulation_zero,
+            axis=tuple(i for i in range(-dim_diff, 0) if i not in skip_axes),
+        )
```
[the remark this answers is not in the document]
Comment (bruno-pannunzio) [reply src/phasorpy/phasor.py]: Thanks @cgohlke ! Your suggestions are excelent, I haven't realized I calculated the axis twice and your way is much more cleaner an efficent.

[the remark this answers is not in the document]
Comment (bruno-pannunzio) [reply src/phasorpy/phasor.py]: Also, I tried and it works for different axis, not just the first one.

**Comment by bruno-pannunzio** [comment]
I will refactor the axis determination code so that it can be used for both `phasor_calibrate` and `phasor_center`. Should this helper function be in the `_utils` module or in the `phasor` module as a hidden function?

**Comment by cgohlke** [comment]
I would prefer the helper function to also validate and "normalize" the `skip_axes` parameter.

Should we rename `skip_axes` to `skip_axis` to match the numpy `axis` argument? 

The helper function should be tested. 

How about something like:

```python
def _parse_skip_axis(
    skip_axis: int | Sequence[int] | None,
    /,
    ndim: int,
) -> tuple[tuple[int, ...], tuple[int, ...]]:
    """Return axes to skip and not to skip.

    This helper function is used to validate and parse `skip_axis`
    parameters.

    Parameters
    ----------
    skip_axis : Sequence of int, or None
        Axes to skip. If None, no axes are skipped.
    ndim : int
        Dimensionality of array in which to skip axes.

    Returns
    -------
    skip_axis
        Ordered, positive values of `skip_axis`.
    other_axis
        Axes indices not included in `skip_axis`.

    Raises
    ------
    IndexError
        If any `skip_axis` value is out of bounds of `ndim`.

    Examples
    --------
    >>> _parse_skip_axis((1, -2), 5)
    (1, 3), (0, 2, 4)

    """
    if ndim < 0:
        raise ValueError(f'invalid {ndim=}')
    if skip_axis is None:
        return (), tuple(range(ndim))
    if not isinstance(skip_axis, Sequence):
        skip_axis = (skip_axis,)
    if any(i >= ndim or i < -ndim for i in skip_axis):
        raise IndexError(f"skip_axis={skip_axis} out of range for {ndim=}")
    skip_axis = tuple(sorted(int(i % ndim) for i in skip_axis))
    other_axis = tuple(i for i in range(ndim) if i not in skip_axis)
    return skip_axis, other_axis
```
and
```python
def test_parse_skip_axis():
    """Test _parse_skip_axis function."""
    assert _parse_skip_axis(None, 0) == ((), ())
    assert _parse_skip_axis(None, 1) == ((), (0,))
    assert _parse_skip_axis((), 1) == ((), (0,))
    assert _parse_skip_axis(0, 1) == ((0,), ())
    assert _parse_skip_axis(0, 2) == ((0,), (1,))
    assert _parse_skip_axis(-1, 2) == ((1,), (0,))
    assert _parse_skip_axis((1, -2), 5) == ((1, 3), (0, 2, 4))
    with pytest.raises(ValueError):
        _parse_skip_axis(0, -1)
    with pytest.raises(IndexError):
        _parse_skip_axis(0, 0)
    with pytest.raises(IndexError):
        _parse_skip_axis(1, 1)
    with pytest.raises(IndexError):
        _parse_skip_axis(-2, 1)
```

**Review by cgohlke — APPROVED** [review]

**Comment by cgohlke** [comment]
Hi @bruno-pannunzio : I think we need to add an example of multi-harmonic calibration to the documentation. How about creating a multi-harmonic version of the introductory tutorial? Or adding it to a later section of the introductory tutorial, for example "Component analysis" if multiple harmonics are used there? Or add an example to the docstring?

If I understand correctly, multi-harmonic calibration can be done by adding the harmonic axis to `skip_axis` and providing an array to `frequency`? Without example, this is not necessarily obvious to library users. In other multi-harmonic-aware functions, we use a `harmonic` parameter, which specifies which harmonics are in the phasor coordinate arrays. Could the `phasor_calibrate` function be made easier to use by adding a `harmonic` parameter?

**Comment by bruno-pannunzio** [comment]
Hi @cgohlke! That is a very nice suggestion and actually you are right that it is not that obvious how to use multi-harmonic calibration. I also think it's a good idea to add the `harmonic` parameter, maybe with it, the `frequency` as array is not needed since it can be calculated from the harmonics and the fundamental frequency. I will try some implementation and open a PR for it. 

Also the addition to the tutorials should be done as you said. I like the idea of a multi-harmonic example that mimics the introductory tutorial, it may also be useful for the component analysis when it's ready.

**Pull request merged**